### ISSUE TITLE
set Fm to eval() mode during reconstruction in backward?

### ISSUE THREAD

**Issue by juntang-zhuang** [silvandeleemput/memcnn#12]
Hi, 

Thanks for your nice work. I wonder during the backward function, the first step is to reconstruct x from y. In this step, should I set Fm and Gm as eval() mode? My concern is, with batchnorm layer, if we reconstruct x from y in train() mode, the running_mean and running_var actually changes, so the reconstructed x is not the actual input x. The same problem happens during construction of a local computation graph in the backward pass. Is there a way to bypass the change in running statistics but still record the gradients?

**Comment by juntang-zhuang** [comment]
Also will the same problem happen with dropout layer?

**Comment by silvandeleemput** [comment]
Hi, thank you for your interest in this project, you might be on to something here. Could you provide a working example of the problem, in particular for batchnorm, and report your python/pytorch versions? This would help to replicate the problem, write suitable tests and see if we can work out a solution. You might also want to provide a little bit more context about what you want to achieve so I'll be able to help you better.

Looking at the source for batchnorm it would indeed appear to receive two updates to the running_mean and running_var instead of one (on forward, and on the reconstruction on backward). However, I don't think that switching to eval mode on the backward pass would properly solve this since batchnorm typically behaves differently on evaluation.  A better strategy might be to write a custom batchnorm class for use inside the Fm and Gm modules which updates the statistics on the first update during training, but which leaves it unchanged after the second update on the forward() method. 

Dropout has indeed a related issue, which is that the dropout pattern is likely to be different on forward() and on the reconstruction on backward, which could complicate the inversion. To combat this a similar strategy as with batchnorm could be employed, where you would write a custom Dropout class for use inside the Fm and Gm modules, which activates and stores a pattern on the first forward call and reactivates the same pattern on the second call during training.

I hope this helps to deal with your problem.

**Comment by juntang-zhuang** [comment]
Hi, I did not run into error, it's just when I looked into the code and performed experiments, the training and validation curves seem unstable. I think batchnorm in eval mode stops changing the running_mean and running_var, but still there are 2 evaluation of batch norm in one backward pass, we can set the 1st as eval, but the 2nd cannot. I agree re-writing batch norm and dropout would be a solution, is it possible for you to implement these layers and push it into the repo? Thanks a lot.

**Comment by silvandeleemput** [comment]
I am sorry I currently do not have the time for that. But, you could give it a go yourself, since it should not be that hard. Just subclass or copy the batchnorm class and add a simple counter which only updates the running statistics every 2 updates. Then in your network code use the modified batchnorm instead of normal batchnorm.

However, it might very well be that it is not necessarily the interplay between the ReversibleBlock and batchnorm which is causing the unstable behavior. So I would suggest you test this first. You can do this, by running your network with the following minimal reversible block nn.module code I have provided below. The nn.module is a drop-in replacement for the ReversibleBlock which performs all the reversible operations but does not perform any memory savings. Additionally, it should call batchnorm only once as normal, since it does not use a custom backward function in which the graph is rebuilt. So, for the test, if after replacing all ReversibleBlocks with the provided ReversibleBlockTest yields more stable training and validation curves (just disable dropout for now), then it was likely caused by issues related to batchnorm in combination with the original ReversibleBlock, otherwise it was caused by something else. 

Let me know if it works or if you need more help.

```python
import copy
import torch
import torch.nn as nn


class ReversibleBlockTest(nn.Module):
    def __init__(self, Fm, Gm=None):
        """

        Parameters
        ----------
            Fm : torch.nn.Module
                A torch.nn.Module encapsulating an arbitrary function

            Gm : torch.nn.Module
                A torch.nn.Module encapsulating an arbitrary function
                (If not specified a deepcopy of Gm is used as a Module)

        """
        super(ReversibleBlockTest, self).__init__()
        # mirror the passed module, without parameter sharing...
        if Gm is None:
            Gm = copy.deepcopy(Fm)
        self.Fm = Fm
        self.Gm = Gm

    def forward(self, x):
        assert (x.shape[1] % 2 == 0)  # assert if possible

        # partition in two equally sized set of channels
        x1, x2 = torch.chunk(x, 2, dim=1)

        # compute outputs from inputs
        y1 = x1 + self.Fm.forward(x2)
        y2 = x2 + self.Gm.forward(y1)

        y = torch.cat([y1, y2], dim=1)
        return y

    def inverse(self, y):
        assert (y.shape[1] % 2 == 0)  # assert if possible

        # partition in two equally sized set of channels
        y1, y2 = torch.chunk(y, 2, dim=1)

        # compute inputs from outputs
        x2 = y2 - self.Gm.forward(y1)
        x1 = y1 - self.Fm.forward(x2)

        x = torch.cat([x1, x2], dim=1)
        return x
```

**Comment by silvandeleemput** [comment]
Closing this issue because it is about an outdated version of MemCNN
